build/toml-patch/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implements patching of TOML documents
+//!
+//! The `toml_edit` crate is great, but it has a major limitation: tables are
+//! ordered with a global `position`.  This means that to insert a new table,
+//! you have to shift everthing after that table downwards, and adjust relative
+//! positions within the new table.
+//!
+//! This crate exposes a single function [`merge_toml_documents`] which does
+//! this for you!
+
+use anyhow::{anyhow, bail, Result};
+use std::collections::BTreeMap;
+use toml_edit::{visit::Visit, visit_mut::VisitMut};
+
+pub fn merge_toml_documents(
+    original: &mut toml_edit::Document,
+    mut patches: toml_edit::Document,
+) -> Result<()> {
+    // Find offsets where we need to insert gaps for incoming patches
+    let mut offsets = BTreeMap::new();
+    compute_offsets(original, &patches, &mut offsets)?;
+    assert!(!offsets.contains_key(&0));
```

Comment:
Probably worth a comment what's going on here (what's special about 0 that we need it to be there but not be put there by `compute_offsets`?).

Reply:
This is for positions _before_ the first offset in the table.  I moved it to an `unwrap_or((0, 0))` instead in be0baefc3d9c8f94dcd8138a55dd7b97244e4fc1